Linux/O_STM/DOXY/man/man3/AIB.3
.TH "AIB" 3 "Sun Apr 1 2018" "C++ Software transactional Memory" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AIB \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <AIB\&.h>\fP
.PP
Inherits \fBBANK\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAIB\fP ()"
.br
.ti -1c
.RI "\fBAIB\fP (int \fBaccountNumber\fP, double \fBbalance\fP, std::string \fBfirstName\fP, std::string \fBlastName\fP, std::string \fBaddress\fP)"
.br
.ti -1c
.RI "\fBAIB\fP (std::shared_ptr< \fBBANK\fP > obj, int _version, int _unique_id)"
.br
.ti -1c
.RI "\fBAIB\fP (const \fBAIB\fP &orig)"
.br
.ti -1c
.RI "virtual void \fBcopy\fP (std::shared_ptr< \fBOSTM\fP > to, std::shared_ptr< \fBOSTM\fP > from)"
.br
.RI "\fIcopy function, make deep copy of the object/pointer \fP"
.ti -1c
.RI "virtual int \fBGetAccountNumber\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetAddress\fP () const "
.br
.ti -1c
.RI "virtual double \fBGetBalance\fP () const "
.br
.ti -1c
.RI "virtual std::shared_ptr< \fBOSTM\fP > \fBgetBaseCopy\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fIgetBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object \fP"
.ti -1c
.RI "virtual std::string \fBGetFirstName\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetFullname\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetLastName\fP () const "
.br
.ti -1c
.RI "\fBAIB\fP \fBoperator=\fP (const \fBAIB\fP &orig)"
.br
.ti -1c
.RI "virtual void \fBSetAccountNumber\fP (int \fBaccountNumber\fP)"
.br
.ti -1c
.RI "virtual void \fBSetAddress\fP (std::string \fBaddress\fP)"
.br
.ti -1c
.RI "virtual void \fBSetBalance\fP (double \fBbalance\fP)"
.br
.ti -1c
.RI "virtual void \fBSetFirstName\fP (std::string \fBfirstName\fP)"
.br
.ti -1c
.RI "virtual void \fBSetFullname\fP (std::string \fBfullname\fP)"
.br
.ti -1c
.RI "virtual void \fBSetLastName\fP (std::string \fBlastName\fP)"
.br
.ti -1c
.RI "virtual void \fBtoString\fP ()"
.br
.RI "\fI_cast, is use to cast bak the std::shared_ptr<OSTM> to the required type \fP"
.ti -1c
.RI "virtual \fB~AIB\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "int \fBaccountNumber\fP"
.br
.ti -1c
.RI "std::string \fBaddress\fP"
.br
.ti -1c
.RI "double \fBbalance\fP"
.br
.ti -1c
.RI "std::string \fBfirstName\fP"
.br
.ti -1c
.RI "std::string \fBfullname\fP"
.br
.ti -1c
.RI "std::string \fBlastName\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Inherit from \fBBANK\fP 
.PP
Definition at line 18 of file AIB\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AIB::AIB ()\fC [inline]\fP"
Constructor 
.PP
Definition at line 23 of file AIB\&.h\&.
.PP
References accountNumber, address, balance, firstName, fullname, and lastName\&.
.PP
Referenced by AIB(), and getBaseCopy()\&.
.PP
.nf
23          : BANK()
24     {
25         this->accountNumber = 0;
26         this->balance = 50;
27         this->firstName = "Joe";
28         this->lastName = "Blog";
29         this->address = "High street, Carlow";
30         this->fullname = firstName + " " + lastName;
31     
32     };
.fi
.SS "AIB::AIB (int accountNumber, double balance, std::string firstName, std::string lastName, std::string address)\fC [inline]\fP"
Custom constructor 
.PP
Definition at line 36 of file AIB\&.h\&.
.PP
References accountNumber, address, balance, firstName, fullname, and lastName\&.
.PP
.nf
36                                                                                                       : BANK()
37     {
38         this->accountNumber = accountNumber;
39         this->balance = balance;
40         this->firstName = firstName;
41         this->lastName = lastName;
42         this->address = address;
43         this->fullname = firstName + " " + lastName;
44     }; 
.fi
.SS "AIB::AIB (std::shared_ptr< \fBBANK\fP > obj, int _version, int _unique_id)\fC [inline]\fP"
Custom constructor, used by the library for deep copying 
.PP
Definition at line 48 of file AIB\&.h\&.
.PP
References accountNumber, address, AIB(), balance, firstName, fullname, and lastName\&.
.PP
.nf
48                                                               : BANK(_version, _unique_id)
49     {
50         
51         this->accountNumber = obj->GetAccountNumber();
52         this->balance = obj->GetBalance();
53         this->firstName = obj->GetFirstName();
54         this->lastName = obj->GetLastName();
55         this->address = obj->GetAddress();
56         this->fullname = obj->GetFirstName() + " " + obj->GetLastName(); 
57         
58     };
.fi
.SS "AIB::AIB (const \fBAIB\fP & orig)"
Copy constructor 
.PP
Definition at line 14 of file AIB\&.cpp\&.
.PP
.nf
14                         {
15 }
.fi
.SS "AIB::~AIB ()\fC [virtual]\fP"
de-constructor 
.PP
Definition at line 17 of file AIB\&.cpp\&.
.PP
Referenced by operator=()\&.
.PP
.nf
17           {
18 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void AIB::copy (std::shared_ptr< \fBOSTM\fP > to, std::shared_ptr< \fBOSTM\fP > from)\fC [virtual]\fP"

.PP
copy function, make deep copy of the object/pointer 
.PP
\fBParameters:\fP
.RS 4
\fIobjTO\fP is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM> 
.br
\fIobjFROM\fP is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM> 
.RE
.PP

.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 37 of file AIB\&.cpp\&.
.PP
References OSTM::Set_Unique_ID()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
37                                                               {
38 
39     std::shared_ptr<AIB> objTO = std::dynamic_pointer_cast<AIB>(to);
40     std::shared_ptr<AIB> objFROM = std::dynamic_pointer_cast<AIB>(from);
41     objTO->Set_Unique_ID(objFROM->Get_Unique_ID());
42     objTO->Set_Version(objFROM->Get_Version());
43     objTO->SetAccountNumber(objFROM->GetAccountNumber());
44     objTO->SetBalance(objFROM->GetBalance());
45 }
.fi
.SS "int AIB::GetAccountNumber () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 81 of file AIB\&.cpp\&.
.PP
References accountNumber\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
81                                 {
82     return accountNumber;
83 }
.fi
.SS "std::string AIB::GetAddress () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 65 of file AIB\&.cpp\&.
.PP
References address\&.
.PP
Referenced by operator=()\&.
.PP
.nf
65                                 {
66     return address;
67 }
.fi
.SS "double AIB::GetBalance () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 73 of file AIB\&.cpp\&.
.PP
References balance\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
73                              {
74     return balance;
75 }
.fi
.SS "std::shared_ptr< \fBOSTM\fP > AIB::getBaseCopy (std::shared_ptr< \fBOSTM\fP > object)\fC [virtual]\fP"

.PP
getBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object 
.PP
\fBParameters:\fP
.RS 4
\fIobjTO\fP is a \fBBANK\fP type pointer for casting 
.br
\fIobj\fP is a std::shared_ptr<BANK> return type 
.RE
.PP

.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 24 of file AIB\&.cpp\&.
.PP
References AIB()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
25 {
26 
27     std::shared_ptr<BANK> objTO = std::dynamic_pointer_cast<BANK>(object);
28     std::shared_ptr<BANK> obj(new AIB(objTO, object->Get_Version(),object->Get_Unique_ID()));
29     std::shared_ptr<OSTM> ostm_obj = std::dynamic_pointer_cast<OSTM>(obj);
30     return ostm_obj;
31 }
.fi
.SS "std::string AIB::GetFirstName () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 97 of file AIB\&.cpp\&.
.PP
References firstName\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
97                                   {
98     return firstName;
99 }
.fi
.SS "std::string AIB::GetFullname () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 105 of file AIB\&.cpp\&.
.PP
References fullname\&.
.PP
Referenced by operator=()\&.
.PP
.nf
105                                  {
106     return fullname;
107 }
.fi
.SS "std::string AIB::GetLastName () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 89 of file AIB\&.cpp\&.
.PP
References lastName\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
89                                  {
90     return lastName;
91 }
.fi
.SS "\fBAIB\fP AIB::operator= (const \fBAIB\fP & orig)\fC [inline]\fP"
Operator 
.PP
Definition at line 66 of file AIB\&.h\&.
.PP
References accountNumber, address, balance, copy(), firstName, fullname, GetAccountNumber(), GetAddress(), GetBalance(), getBaseCopy(), GetFirstName(), GetFullname(), GetLastName(), lastName, SetAccountNumber(), SetAddress(), SetBalance(), SetFirstName(), SetFullname(), SetLastName(), toString(), and ~AIB()\&.
.PP
.nf
66 {};
.fi
.SS "void AIB::SetAccountNumber (int accountNumber)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 77 of file AIB\&.cpp\&.
.PP
References accountNumber\&.
.PP
Referenced by operator=()\&.
.PP
.nf
77                                             {
78     this->accountNumber = accountNumber;
79 }
.fi
.SS "void AIB::SetAddress (std::string address)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 61 of file AIB\&.cpp\&.
.PP
References address\&.
.PP
Referenced by operator=()\&.
.PP
.nf
61                                       {
62     this->address = address;
63 }
.fi
.SS "void AIB::SetBalance (double balance)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 69 of file AIB\&.cpp\&.
.PP
References balance\&.
.PP
Referenced by operator=()\&.
.PP
.nf
69                                    {
70     this->balance = balance;
71 }
.fi
.SS "void AIB::SetFirstName (std::string firstName)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 93 of file AIB\&.cpp\&.
.PP
References firstName\&.
.PP
Referenced by operator=()\&.
.PP
.nf
93                                           {
94     this->firstName = firstName;
95 }
.fi
.SS "void AIB::SetFullname (std::string fullname)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 101 of file AIB\&.cpp\&.
.PP
References fullname\&.
.PP
Referenced by operator=()\&.
.PP
.nf
101                                         {
102     this->fullname = fullname;
103 }
.fi
.SS "void AIB::SetLastName (std::string lastName)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 85 of file AIB\&.cpp\&.
.PP
References lastName\&.
.PP
Referenced by operator=()\&.
.PP
.nf
85                                         {
86     this->lastName = lastName;
87 }
.fi
.SS "void AIB::toString ()\fC [virtual]\fP"

.PP
_cast, is use to cast bak the std::shared_ptr<OSTM> to the required type toString function, displays the object values in formatted way 
.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 56 of file AIB\&.cpp\&.
.PP
References OSTM::Get_Unique_ID(), OSTM::Get_Version(), GetAccountNumber(), GetBalance(), GetFirstName(), and GetLastName()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
57 {
58     std::cout << "\nAIB BANK" << "\nUnique ID : " << this->Get_Unique_ID() << "\nInt account : " << this->GetAccountNumber() << "\nDouble value : " << this->GetBalance() << "\nFirst name: " << this->GetFirstName() << "\nLast name : " << this->GetLastName()  << "\nVersion number : " << this->Get_Version() << std::endl;
59 }
.fi
.SH "Member Data Documentation"
.PP 
.SS "int AIB::accountNumber\fC [private]\fP"

.PP
Definition at line 100 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetAccountNumber(), operator=(), and SetAccountNumber()\&.
.SS "std::string AIB::address\fC [private]\fP"

.PP
Definition at line 102 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetAddress(), operator=(), and SetAddress()\&.
.SS "double AIB::balance\fC [private]\fP"

.PP
Definition at line 101 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetBalance(), operator=(), and SetBalance()\&.
.SS "std::string AIB::firstName\fC [private]\fP"

.PP
Definition at line 98 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetFirstName(), operator=(), and SetFirstName()\&.
.SS "std::string AIB::fullname\fC [private]\fP"

.PP
Definition at line 97 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetFullname(), operator=(), and SetFullname()\&.
.SS "std::string AIB::lastName\fC [private]\fP"

.PP
Definition at line 99 of file AIB\&.h\&.
.PP
Referenced by AIB(), GetLastName(), operator=(), and SetLastName()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for C++ Software transactional Memory from the source code\&.


Linux/SAMPLE_APP/AIB.cpp
/* 
 * File:   AIB.cpp
 * Author: Zoltan Fuzesi C00197361, 
 * IT Carlow, Software Engineering, 
 *
 * Supervisor : Joe Kehoe, 
 *
 * C++ Software Transactional Memory, 
 * 
 * Created on December 18, 2017, 2:09 PM
 * OSTM base class function declarations.
 */

#include <math.h>

#include "AIB.h"


AIB::~AIB() {
}
/*!
 * \brief getBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object
 * \param object is a OSTM type shared pointer use to create a new copy of the pointer
 */
std::shared_ptr<OSTM> AIB::getBaseCopy(std::shared_ptr<OSTM> object)
{
    /** Dynamic cast from OSTM to BANK type*/
    std::shared_ptr<BANK> objTO = std::dynamic_pointer_cast<BANK>(object);
    /** BANK type Instance creation shared pointer*/
    std::shared_ptr<BANK> obj(new AIB(objTO, object->Get_Version(),object->Get_Unique_ID()));
    /** Dynamic cast from BANK to OSTM type*/
    std::shared_ptr<OSTM> ostm_obj = std::dynamic_pointer_cast<OSTM>(obj);
    /** Return new OSTM copy onject*/
    return ostm_obj;
}
/*!
 * \brief copy function, make deep copy of the object/pointer
 * \param to std::shared_ptr<OSTM>, BANK type shared pointer used to copy into the values
 * \param from std::shared_ptr<OSTM>, BANK type shared ponter used to get object values tfrom copy
 */
void AIB::copy(std::shared_ptr<OSTM> to, std::shared_ptr<OSTM> from){

    /** Dynamic cast from OSTM to AIB*/
    std::shared_ptr<AIB> objTO = std::dynamic_pointer_cast<AIB>(to);
    /** Dynamic cast from OSTM to AIB */
    std::shared_ptr<AIB> objFROM = std::dynamic_pointer_cast<AIB>(from);
    /** Set values fro object to object*/
    objTO->Set_Unique_ID(objFROM->Get_Unique_ID());
    /** Set values fro object to object*/
    objTO->Set_Version(objFROM->Get_Version());
    /** Set values fro object to object*/
    objTO->SetAccountNumber(objFROM->GetAccountNumber());
    /** Set values fro object to object*/
    objTO->SetBalance(objFROM->GetBalance());
}
/*!
 *  \brief toString function, displays the object values in formatted way
 */
void AIB::toString()
{
    std::cout << "\nAIB BANK" << "\nUnique ID : " << this->Get_Unique_ID() << "\nInt account : " << this->GetAccountNumber() << "\nDouble value : " << this->GetBalance() << "\nFirst name: " << this->GetFirstName() << "\nLast name : " << this->GetLastName()  << "\nVersion number : " << this->Get_Version() << std::endl;
}
/*!
 *  \brief SetAddress setter for address private field
 */
void AIB::SetAddress(std::string address) {
    this->address = address;
}
/*!
 *  \brief GetAddress getter for address private field
 */
std::string AIB::GetAddress() const {
    return address;
}
/*!
 *  \brief SetBalance setter for balance private field
 */
void AIB::SetBalance(double balance) {
    this->balance = balance;
}
/*!
 *  \brief GetBalance getter for balance private field
 */
double AIB::GetBalance() const {
    return balance;
}
/*!
 *  \brief SetAccountNumber setter for accountNumber private field
 */
void AIB::SetAccountNumber(int accountNumber) {
    this->accountNumber = accountNumber;
}
/*!
 *  \brief GetAccountNumber getter for accountNumber private field
 */
int AIB::GetAccountNumber() const {
    return accountNumber;
}
/*!
 *  \brief SetLastName setter for lastName private field
 */
void AIB::SetLastName(std::string lastName) {
    this->lastName = lastName;
}
/*!
 *  \brief GetLastName getter for lastName private field
 */
std::string AIB::GetLastName() const {
    return lastName;
}
/*!
 *  \brief SetFirstName setter for firstName private field
 */
void AIB::SetFirstName(std::string firstName) {
    this->firstName = firstName;
}
/*!
 *  \brief GetFirstName getter for firstName private field
 */
std::string AIB::GetFirstName() const {
    return firstName;
}
/*!
 *  \brief SetFullname setter for fullname private field
 */
void AIB::SetFullname(std::string fullname) {
    this->fullname = fullname;
}
/*!
 *  \brief GetFullname getter for fullname private field
 */
std::string AIB::GetFullname() const {
    return fullname;
}
